infra/notifications/src/app.js

```diff
@@ -165,68 +165,107 @@ app.post('/', async (req, res) => {
 
 /**
  * Endpoint called from the mobile marketplace app to register a device token
- * and Ethereum address.
+ * and Ethereum address for push notifications.
+ *
+ * Notes:
+ *  - The app may call this end point more than once.
+ *  - On iOS devices, the app may call this endpoint before the user accepts
+ *    to turn on notifications. So the deviceToken argument may be empty.
  */
 app.post('/mobile/register', async (req, res) => {
-  logger.info('Call to mobile device registry endpoint')
-
   const mobileRegister = {
-    ethAddress: req.body.eth_address,
-    deviceType: req.body.device_type,
+    ethAddress: _.get(req.body, 'eth_address', null),
+    deviceType: _.get(req.body, 'device_type', null),
     deviceToken: _.get(req.body, 'device_token', null),
-    permissions: req.body.permissions
+    permissions: _.get(req.body, 'permissions', null)
   }
+  logger.info(`POST /mobile/register for ${mobileRegister.ethAddress}`)
 
-  // See if a row already exists for this device/address
-  let registryRow = await MobileRegistry.findOne({
-    where: {
-      ethAddress: mobileRegister.ethAddress,
-      deviceToken: mobileRegister.deviceToken
+  if (!mobileRegister.ethAddress) {
+    return res.status(400).send({ errors: ['Invalid Eth address'] })
+  }
+  // By convention all Eth addresses are stored lower cased in the DB.
+  mobileRegister.ethAddress = mobileRegister.ethAddress.toLowerCase()
+
+  if (mobileRegister.deviceToken) {
+    // See if a row already exists for this (Eth address, device token) pair.
+    const registryRow = await MobileRegistry.findOne({
+      where: {
+        ethAddress: mobileRegister.ethAddress,
+        deviceToken: mobileRegister.deviceToken
+      }
+    })
+
+    if (!registryRow) {
+      // Nothing exists, create a new row.
+      await MobileRegistry.create(mobileRegister)
+      res.sendStatus(201)
+      logger.debug('Added new mobile device to registry: ', req.body)
+    } else {
+      // Update existing row since permissions might have changed.
+      await MobileRegistry.upsert(mobileRegister)
```

Comment:
if you're doing an upsert anyway, is there any added benefit to the insert above?

Reply:
Good catch. I blindly copied the existing code. Does not make sense. Fixed to do an update if permissions change.